### PR TITLE
Use internal SYCL kernel range rounding for RangePolicy ParallelFor

### PR DESCRIPTION
**Pull request by masterleinad**
Simplified version of #4405. We need to provide a kernel name for SYCL to do reasonable kernel range rounding internally.

### CODE REVIEW

**Comment by masterleinad** [comment]
Without this, the workgroup size is chosen as a divisor of the workgroup range leading to poor performance if, e.g., the range is a prime number.
For a simple add kernel,
```
n = 524287
AXPBY KK: 8.868529e-01 s 1.321383e+01 GB/s
n = 524288
AXPBY KK: 4.283028e-02 s 2.736090e+02 GB/s
n = 524289
AXPBY KK: 3.133845e-01 s 3.739423e+01 GB/s
```
before this patch
```
n = 524287
AXPBY KK: 4.648326e-02 s 2.521064e+02 GB/s
n = 524288
AXPBY KK: 4.585346e-02 s 2.555696e+02 GB/s
n = 524289
AXPBY KK: 4.766056e-02 s 2.458799e+02 GB/s
```
after this patch.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp [hunk 1]
```diff
@@ -72,17 +90,12 @@ class Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>,
...
-      const auto begin = policy.begin();
-
-      cgh.parallel_for(range, [=](sycl::item<1> item) {
-        const typename Policy::index_type id = item.get_linear_id() + begin;
-        if constexpr (std::is_same<WorkTag, void>::value)
-          functor(id);
-        else
-          functor(WorkTag(), id);
-      });
+      cgh.parallel_for<FunctorWrapperRangePolicyParallelFor<Functor, Policy>>(
```
Comment (masterleinad) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:97]: The crucial change is being able to provide the template argument, `FunctorWrapperRangePolicyParallelFor<Functor, Policy>`, here (and I couldn't get that to work with the lambda). 
Using a functor also matches the change for https://github.com/kokkos/kokkos/pull/4097.

**Review by dalg24 — COMMENTED** [review]
I am fine with the changes proposed in this PR but I fail to see how they relate to the title line

**Comment by masterleinad** [comment]
> I am fine with the changes proposed in this PR but I fail to see how they relate to the title line

I changed the title.

**Review by dalg24 — APPROVED** [review]
"kernel rounding"?

**Review by Rombur — APPROVED** [review]

**Pull request merged**